File: Builder/C#/BuilderPattern/BuilderPattern/Builder/Builder.cs
namespace BuilderPattern
{
    class Builder : AbstractBuilder
    {
        private House house;

        public Builder()
        {
            house = new House();
        }
        public override void BuildBasement()
        {
            house.Add(new Basement());
        }

        public override void BuildStorey()
        {
            house.Add(new Storey());
        }

        public override void BuildRoof()
        {
            house.Add(new Roof());
        }

        public override House GetResult()
        {
            return house;
        }
    }
}

File: Builder/C++/BuilderPattern/BuilderPattern/Builder.cpp
#include "Builder.h"
#include "IHousePart.h"

void Builder::BuildBasement()
{
	house.Add((IHousePart*)new Basement());
}

void Builder::BuildStorey()
{
	house.Add((IHousePart*)new Storey());
}

void Builder::BuildRoof()
{
	house.Add((IHousePart*)new Roof());
}

House Builder::GetResult()
{
	return house;
}
